### PR TITLE
[1.0] make determining when a given stage is finished much more reliable

### PR DESCRIPTION
**Pull request by KyleAMathews**
Two main tricks.

One was adding a way to track cascading work e.g. a source plugin adds a node which then transformer plugin takes and creates a new node which then another transformer plugin does more work on and so forth. Before there was no way to know when the last bit of work associated with the initial sourcing was finished. To solve this, I added a `traceId` which is passed in during the original API call and then all subsequent work passes it along. And rather nicely, this is all hidden from the plugins themselves as we inject the traceId into the actions that plugins create.

Second was moving more internal work into internal plugins. This lets us take advantage of the new tracing capabilities above so that if internal work is triggered by an API call, we're tracking it along with every other plugin.

Now the bootstrap/index.js file is quite straightforward to read especially with our use of async/await. It's now stepping cleanly through each step of the process.

This solves two problems. The biggest is there was occasionally weird graphql errors where some data just wouldn't get processed in time before the debounce timeout. Those should be gone now. The other is the debouncing added time overhead to the bootstrap process. Removing that means bootstrap is sped up by something like 750-1000 milliseconds. Now a barebones Gatsby site's bootstrap can be as fast as 2.5-3 seconds (And most of that time is loading JavaScript so can't wait for prepack.io to get prod ready).

I also snuck in a simple jobs system. gatsby-plugin-sharp (image processing plugin) is the only user atm but it let's plugins tell Gatsby it has long-running jobs going and block bootstrap from finishing while it's working. It should have all the data necessary for building a nice CLI UI @jquense

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Oh and I also added support to the API runner for callback based async. So you can now either return a promise or use the optional third callback argument.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 793119b69503f7a4d2d136eaec1fb85f145784ff

https://app.netlify.com/sites/using-postcss-sass/deploys/59303449cf321c04c3ac20e3

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 793119b69503f7a4d2d136eaec1fb85f145784ff

https://app.netlify.com/sites/image-processing/deploys/59303449cf321c04c3ac20e8

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 793119b69503f7a4d2d136eaec1fb85f145784ff

https://app.netlify.com/sites/gatsbygram/deploys/59303448cf321c04c3ac20dc

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 793119b69503f7a4d2d136eaec1fb85f145784ff

https://app.netlify.com/sites/gatsbyjs/deploys/59303448cf321c04c3ac20da

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit d30369c1bc398e85ff6c906c0c986e8536b235b0

https://app.netlify.com/sites/using-postcss-sass/deploys/5930360f6f4c506b0580058d

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit d30369c1bc398e85ff6c906c0c986e8536b235b0

https://app.netlify.com/sites/gatsbyjs/deploys/5930360f6f4c506b05800585

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit d30369c1bc398e85ff6c906c0c986e8536b235b0

https://app.netlify.com/sites/gatsbygram/deploys/5930360f6f4c506b05800587

**Comment by jquense** [comment]
this is sweet! Should make wrangling console output a lot more straightforward as well :) I can't wait to try on top of this

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit d30369c1bc398e85ff6c906c0c986e8536b235b0

https://app.netlify.com/sites/image-processing/deploys/5930360f6f4c506b05800591

File: packages/gatsby-plugin-sharp/src/index.js [hunk 1]
```diff
@@ -10,7 +10,15 @@ const imageminPngquant = require(`imagemin-pngquant`)
 const queue = require(`async/queue`)
 const duotone = require(`./duotone`)
 
-const { boundActionCreators } = require(`gatsby/dist/redux/actions`)
+// This fails in tests.
+let actions
+try {
+  actions = require(`gatsby/dist/redux/actions`)
```
Comment (jquense) [inline packages/gatsby-plugin-sharp/src/index.js]: could use Jest moduleMapper to rewrite these paths back to `src` in tests

Comment (KyleAMathews) [inline packages/gatsby-plugin-sharp/src/index.js]: 👍 done

**Comment by KyleAMathews** [comment]
Things are finally all passing but now mysteriously, one of the function names isn't getting parsed. But only on Linux apparently as it works just fine on my mac. Oy!

screen shot 2017-06-01 at 11 41 21 am

Close enough. We'll figure that out another time.

**Pull request merged**